lib/semantics/check-do.cc

```diff
@@ -219,6 +214,46 @@ class DoConcurrentBodyEnforce {
   }
 
 private:
+  void CheckAllocateObjectIsCoarray(
+      const parser::AllocateObject &allocateObject, const char *stmtType) {
+    const Symbol *symbol{nullptr};
+    parser::CharBlock position;
+    std::visit(
+        common::visitors{
+            [&](const parser::Name &x) {
+              symbol = x.symbol;
+              position = x.source;
+            },
+            [&](const parser::StructureComponent &x) {
+              symbol = x.component.symbol;
+              position = x.component.source;
+            },
+        },
+        allocateObject.u);
+    if (IsCoarray(*symbol)) {
+      context_.Say(position,
+          "%s coarray not allowed in DO CONCURRENT"_err_en_US, stmtType);
+    }
+  }
+
+  void CheckAllocateContainsCoarray(const parser::AllocateStmt &allocateStmt) {
+    const std::list<parser::Allocation> &allocationList{
+        std::get<1>(allocateStmt.t)};
+    for (const parser::Allocation &allocation : allocationList) {
```

Comment:
you can use auto here and in the loop in `CheckDeallocateContainsCoarray()`

```suggestion
    for (auto &allocation : allocationList) {
```

Reply:
We like to be explicitly `const` though.

Reply:
Right, I've updated by suggestion.

Reply:
Varun's suggestion has a ```const``` in it.

I prefer not to use ```auto``` in this case because I think it's slightly more readable to explicitly state the type.  But in this case I could go either way since the type of ```allocation``` is pretty obvious.